src/chain/shared_state.rs

```diff
@@ -537,8 +543,111 @@ impl SectionProofChain {
         self.blocks.len()
     }
 
+    pub fn last_prefix_version(&self) -> (&Prefix<XorName>, u64) {
+        self.blocks
+            .last()
+            .map(|block| (block.prefix(), block.version()))
+            .unwrap_or((&self.prefix, self.version))
+    }
+
+    #[cfg(all(test, feature = "mock"))]
+    pub fn all_prefix_version(&self) -> impl DoubleEndedIterator<Item = (&Prefix<XorName>, u64)> {
+        iter::once((&self.prefix, self.version)).chain(
+            self.blocks
+                .iter()
+                .map(|block| (block.prefix(), block.version())),
+        )
+    }
+
+    pub fn last_new_public_key_info(&self) -> Option<&SectionKeyInfo> {
+        self.blocks.last().map(|block| block.key_info())
+    }
+
+    fn last_trusted_key_info<'a>(
+        &'a self,
+        last_trusted: &'a SectionKeyInfo,
+    ) -> Option<&'a SectionKeyInfo> {
+        let block_offset = last_trusted.version().saturating_sub(self.version) as usize;
+
+        if block_offset == 0 {
```

Comment:
basically, this block of code is saying:
we only consider two cases are valid:
  a, last_trusted equals to self
  b, the block at (block_offset - 1) contains the same key_info as last_trusted
 
right?

Reply:
Yes.
Really there is only one case:
 -  the block at (block_offset - 1) contains the same key_info as last_trusted.
It is just that in the case of block_offset, we only have prefix and version for `block_offset - 1` and they are self.version, self.preffix.

Reply:
ok. thx for the clarification